server/gui/utils.py

```diff
@@ -13,13 +13,49 @@ class WorkerSignals(QObject):
     Defines the signals available from a running worker thread.
     Supported signals are:
         finished
+        ready
         error - `str` error message
         result - `object` data returned from processing, anything
     """
     finished = Signal()
     error = Signal(str)
     result = Signal(object)
+    ready = Signal()
+
+
+class SiteReadySignals(QObject):
+    """
+    Defines the signals available from a running worker thread.
+    Supported signals are:
+        timeout
+        ready
+        error - `str` error message
+    """
+    ready = Signal()
+    timeout = Signal()
+    error = Signal(str)
 
 
 class FileLoadSignals(QObject):
     selectedFile = Signal(str)
+
+
+class Emitter:
+    def __init__(self, transport, signals):
+        self.transport = transport
+        self.signals = signals()
+
+    def _emit(self, signature, args=None):
+        if args:
+            getattr(self.signals, signature).emit(args)
+        else:
+            getattr(self.signals, signature).emit()
+
+    def run(self):
+        while True:
+            try:
+                signature = self.transport.recv()
+            except Exception:
+                break
```

Comment:
I assume that the silent break out of the loop is to handle pipe EOF (eg, if the child process exits, the pipe is closed and you expect an EOF).  However, this will not occur because both processes keep both ends of the pipe open.

General best practices for pipes:  after the fork, the parent should close the child end of the pipe, and the child should close the parent end.   If both processes have both ends open, you will not get an EOF and your emitter won't terminate.

Reply:
Additional thought on the silent break -- unless you watch for the emitter thread exit (which I don't see in the code), this silent break could result in the HTTP server dying (app no longer functional), and the UI not noticing.    Suggest that the exception catch at least emit an error signal, so the GUI knows something went wrong and can inform the user (and then either bail/quit, or restart the Worker).

Reply:
Good feedback. Will fix emitter.